src/main/java/com/ldtteam/structures/client/BlueprintHandler.java

```diff
@@ -62,26 +52,49 @@ public static BlueprintHandler getInstance()
      * @param blueprint the wayPointBlueprint to draw.
      * @param pos       its position.
      */
-    public void draw(
-        final Blueprint blueprint,
-        final BlockPos pos,
-        final MatrixStack stack,
-        final float partialTicks)
+    public void draw(final Blueprint blueprint, final BlockPos pos, final MatrixStack stack, final float partialTicks)
     {
         if (blueprint == null)
         {
             Log.getLogger().warn("Trying to draw null blueprint!");
             return;
         }
+        Minecraft.getInstance().getProfiler().startSection("struct_render_cache");
 
-        try
+        final int blueprintHash = blueprint.hashCode();
+        final SoftReference<BlueprintRenderer> rendererRef = rendererCache.get(blueprintHash);
+        final BlueprintRenderer renderer;
+        final long now = System.nanoTime();
+
+        if (rendererRef == null || rendererRef.get() == null)
+        {
+            renderer = BlueprintRenderer.buildRendererForBlueprint(blueprint);
+            rendererCache.put(blueprintHash, new SoftReference<>(renderer));
+        }
+        else
         {
-            blueprintBufferBuilderCache.get(blueprint, () -> BlueprintRenderer.buildRendererForBlueprint(blueprint)).draw(pos, stack, partialTicks);
+            renderer = rendererRef.get();
         }
-        catch (ExecutionException e)
+
+        evictTimeCache.put(blueprintHash, now);
+        renderer.updateBlueprint(blueprint);
+        renderer.draw(pos, stack, partialTicks);
+
+        for (final Int2LongMap.Entry entry : evictTimeCache.int2LongEntrySet())
```

Comment:
I think this is probably not a good idea, to replicate cache-like behaviour without using a cache and looping over it each draw call, which are very high frequency probably gonna cause quite some fps loss.

Reply:
The problem with old mechanism was that it simply haven't worked as cache, according to the old cache javadoc something like this is done every access so I doubt it's slower than the old one

Reply:
Regarding fps loss, this entire patch is just for gigantic structures (more caching in blueprint class) and more lagless build tool usage (renderer refresh was called unnecessarily)

Reply:
the old one didnt have a timeout, just a maximum elements and caches aren't doing that stuff each access, else they'd be quite slow/unusable. Old one was a guava cache, see https://github.com/google/guava/wiki/CachesExplained#timed-eviction for how it works. Granted there are other caches out there aswell, some are faster. see here for example: https://cruftex.net/2016/03/16/Java-Caching-Benchmarks-2016-Part-1.html

Reply:
as I said various builder options (incl. timed eviction) wasn't successful
I can probably connect this to everyGameSecond event instead of this one

Reply:
ye the other option to keep this would be to do it on a tick event, like simply subscribing to the client tick event and evaluate the times there. Should still be a lot better than here